### PR TITLE
feat(): add support for DateOnly and TimeOnly #680

### LINKED ISSUE #680: [FEATURE] Serialize DateOnly as a date instead of serializing its properties individually
### Is your feature request related to a problem?
DateOnly properties are treated as any other complex type and its properties are serialised individually. It prevents using features like date ranges when searching.

### What solution would you like?
DateOnly properties be treated as DateTime properties, except maybe they'd use a different format (like `yyyy-mm-dd`)

### What alternatives have you considered?
Customizing the serialization behavior myself, but I didn't find a way to do so. Also tried applying [DateAttribute] to my property but didn't seem to change anything

### Do you have any additional context?
Currently [DateOnly](https://learn.microsoft.com/en-us/dotnet/api/system.dateonly) properties are serialized this way :
```json
"dateOnlyProperty": {
    "year": 2024,
    "month": 4,
    "day": 15,
    "dayOfWeek": 1,
    "dayOfYear": 106,
    "dayNumber": 738990
},
```

PS : `TimeOnly` may have the same kind of issues

### PR DESCRIPTION
**Pull request by joefeser**
### Description
Add support for DateOnly and TimeOnly that were both added in .net 6

### Issues Resolved
Closes #680.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: global.json [hunk 1]
```diff
@@ -1,10 +1,10 @@
 {
   "sdk": {
-    "version": "6.0.403",
+    "version": "8.0.107",
```
Comment (Xtansia) [inline global.json:3]: The github workflow `setup-dotnet` steps can be updated to change the sdk versions being installed.

File: src/OpenSearch.Net/Transport/PostData.cs [hunk 2]
```diff
@@ -85,8 +85,8 @@ public abstract class PostData
 
 		public static PostData Bytes(byte[] bytes) => new PostData<object>(bytes);
 
-#if NETSTANDARD2_1
-		public static PostData ReadOnlyMemory(ReadOnlyMemory<byte> bytes) => new PostData<object>(bytes);
+#if NETSTANDARD2_1 || NET6_0_OR_GREATER
+		public static PostData ReadOnlyMemory(ReadOnlyMemory<byte> bytes) => new PostData<object>(bytes.ToArray());
```
Comment (Xtansia) [inline src/OpenSearch.Net/Transport/PostData.cs:89]: Please remove the `.ToArray()`, this should be using the constructor here: https://github.com/opensearch-project/opensearch-net/blob/main/src/OpenSearch.Net/Transport/PostData.cs#L149

You'll just need to update the `#if`s further down.

File: tests/Tests/ClientConcepts/LowLevel/PostData.doc.cs [hunk 3]
```diff
@@ -1,3 +1,5 @@
+#pragma warning disable IDE1006 // Naming Styles
+#pragma warning disable IDE0044 // Add readonly modifier
```
Comment (Xtansia) [inline tests/Tests/ClientConcepts/LowLevel/PostData.doc.cs:2]: Please move these `#pragma`s below the license header

**Review by Xtansia — CHANGES_REQUESTED** [review]
Thanks for tackling this @joefeser!
Some relatively small issues other than those already commented in particular files:
- Looks like there's a compile error with the change in SDK.
- Please add an entry to the CHANGELOG.md under the `### Added` section
- The DCO check requires commits to be signed off, you can do this for your existing commits via `git rebase HEAD~2 --signoff` and then make sure to pass `--signoff` to any future commits for this repo

**Comment by joefeser** [comment]
Working on the changes as we speak. Thank you for being so proactive on my PR

File: global.json (hunk 1 above)
Comment (joefeser) [reply global.json:3]: Resolved

File: tests/Tests/ClientConcepts/LowLevel/PostData.doc.cs (hunk 3 above)
Comment (joefeser) [reply tests/Tests/ClientConcepts/LowLevel/PostData.doc.cs]: Resolved

File: src/OpenSearch.Net/Transport/PostData.cs (hunk 2 above)
Comment (joefeser) [reply src/OpenSearch.Net/Transport/PostData.cs]: I think I understood this comment, it is in this commit bcbd1a2161d263ecb010daf522896965ba39531f

**Comment by joefeser** [comment]
> Thanks for tackling this @joefeser! Some relatively small issues other than those already commented in particular files:
> 
> * Looks like there's a compile error with the change in SDK.
> * Please add an entry to the CHANGELOG.md under the `### Added` section
> * The DCO check requires commits to be signed off, you can do this for your existing commits via `git rebase HEAD~2 --signoff` and then make sure to pass `--signoff` to any future commits for this repo

All of these tasks are completed

**Review by Xtansia — APPROVED** [review]

**Comment by Xtansia** [comment]
Thank you for this @joefeser, would you mind making a follow up PR adding the DateOnly and TimeOnly types into the unit test here: https://github.com/opensearch-project/opensearch-net/blob/main/tests/Tests.Reproduce/DateSerialization.cs

**Comment by joefeser** [comment]
@Xtansia I will add it in the next few days. I wasn't sure where to write the tests. I am going to do a POC also to add NodaTime support and at least get that conversation going

**Comment by nalka0** [comment]
I have played a bit with this PR and I think there we still have some likely problems :
- If you don't do anything to the mappings, your `TimeOnly` fields are mapped as strings with
```json
"type": "text",
"fields": {
    "keyword": {
        "type": "keyword",
        "ignore_above": 256
    }
}
```
instead of something with `"type": "date"`
- If you use `openSearchClient.Indices.CreateAsync(indexName, x => x.Map(x => x.AutoMap(typeof(YourType))))` your `DateOnly` and `TimeOnly` properties have no way to be inferred and will be still be treated as "normal" objects here. https://github.com/opensearch-project/opensearch-net/blob/c2fdd5d1df4929000249d07ccdf2c8b21fd2fde6/src/OpenSearch.Client/Mapping/Visitor/PropertyWalker.cs#L197

I tried fixing this second issue and for `DateOnly`, it seems like adding a `case "DateOnly"` next to the DateTime one above on the method I linked is sufficient.

On the other hand for `TimeOnly` it seems like we need more than that since you get a parsing error when trying to put data in your index. From my understanding, the format is responsible there but adding
```csharp
case "TimeOnly":
    return new DateProperty()
    {
        Format = "strict_hour_minute_second_millis"
    };
```
was unsuccesful (for some reasons, doing `Format = ...` there doesn't seem to change anything to the resulting mappings).

PS : I pulled the value for `Format = ` [here](https://opensearch.org/docs/latest/field-types/supported-field-types/date/#formats) but am unsure if `strict_hour_minute_second_millis` is really the best fit

**Pull request merged**